In [3]:
a, b = map(int, input().split())
print(a, b)
print(a+b)

34 342
376


In [4]:
arr = [10, 11, 12, 13, 14]
arr[2]

12

In [5]:
arr.append(15)
arr

[10, 11, 12, 13, 14, 15]

**연결리스트**

클래스를 사용하여 노드를 만들고 이들 노드를 서로 연결해주는 방식을 사용합니다. 각 노드는 데이터와 다음 노드를 가리키는 포인터를 가지고 있습니다. 이렇게 노드들이 서로 연결되면 연결리스트가 만들어집니다.

In [6]:
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def add_node(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current_node = self.head
            while current_node.next is not None:
                current_node = current_node.next
            current_node.next = new_node

    def print_list(self):
        current_node = self.head
        while current_node is not None:
            print(current_node.data)
            current_node = current_node.next

In [7]:
linked_list = LinkedList()

linked_list.add_node(1)
linked_list.add_node(2)
linked_list.add_node(3)

linked_list.print_list()

1
2
3


**스택**

먼저 들어간게 제일 나중에 나가고 (FILO), 나중에 들어간게 제일 먼저 나간다 (LIFO) (ex. 웹브라우저 뒤로가기)


In [8]:
stack = []

# push 연산
stack.append(1)
stack.append(2)
stack.append(3)

# pop 연산
top = stack.pop()  # top 변수에는 3이 저장됩니다.
print(stack)
print(top)

[1, 2]
3


In [9]:
#근데 이게 굳이 의미가 없는데 파이썬은 리스트가 강력해
s = []
s.append(1)
s.append(2)
s.append(3)

t = s[-1]

print(s)
print("마지막에 들어간 3이 출력된다 :", t)

[1, 2, 3]
마지막에 들어간 3이 출력된다 : 3


**큐**

처음에 들어간게 처음나가고 (FIFO) 나중에 들어간게 나중에 나간다 (LILO) (ex. 운영체제에서의 작업리스트)

In [10]:
import queue

q = queue.Queue()

for i in range(1,11):
    q.put(i)

print("처음들어간 순서대로 출력된다")
while not q.empty():
    print(q.get())

처음들어간 순서대로 출력된다
1
2
3
4
5
6
7
8
9
10


**우선순위 큐 Priority Queue (Heap)**

힙을 이용한 구현에서는 삭제 연산과 삽입 연산 모두 O(log n)의 시간 복잡도를 가지므로, 배열을 이용한 구현보다 효율적입니다. 또한, 힙은 이진트리이므로 삽입과 삭제가 일어나도 항상 완전 이진트리의 형태를 유지하기 때문에, 삽입과 삭제 연산에서도 데이터의 이동이 최소화되어 성능이 우수합니다. 최상단 루트에는 항상 최소값이 위치하게 됩니다.


In [11]:
import heapq

# 빈 리스트 생성
pq = []

# 값 추가
heapq.heappush(pq, 312)
heapq.heappush(pq, 1)
heapq.heappush(pq, 33)
heapq.heappush(pq, 16)
heapq.heappush(pq, -10)

print(pq)

print("최상단 노드에는 항상 가장 최소값이 위치하게 됨 :", pq[0])

[-10, 1, 33, 312, 16]
최상단 노드에는 항상 가장 최소값이 위치하게 됨 : -10


**맵 Dictionary**

Key와 Value값을 가지는 자료구조, C++에선 red-black tree구조로 구현되어있으며 O(logN) 파이썬에서는 hash로 구현되어 있다 O(1)

In [12]:
m = {}
m['KIM'] = 100
m['LEE'] = 40
m['PARK'] = 60

print("size :", len(m))
print(m)

size : 3
{'KIM': 100, 'LEE': 40, 'PARK': 60}


**집합 Set**

중복을 허용하지 않는다. 파이썬에서 마찬가지로 hash로 구현되어 있기때문에 O(1)

In [13]:
s = set()

s.add(10)
s.add(12)
s.add(15)
s.add(10)
s.add(17)
s.add(10)

print("10이 중복해서 입력했지만 중복값을 없애버렸다 :",s)

10이 중복해서 입력했지만 중복값을 없애버렸다 : {17, 10, 12, 15}


**9012. 괄호**

괄호 문자열(Parenthesis String, PS)은 두 개의 괄호 기호인 ‘(’ 와 ‘)’ 만으로 구성되어 있는 문자열이다. 그 중에서 괄호의 모양이 바르게 구성된 문자열을 올바른 괄호 문자열(Valid PS, VPS)이라고 부른다. 한 쌍의 괄호 기호로 된 “( )” 문자열은 기본 VPS 이라고 부른다. 만일 x 가 VPS 라면 이것을 하나의 괄호에 넣은 새로운 문자열 “(x)”도 VPS 가 된다. 그리고 두 VPS x 와 y를 접합(concatenation)시킨 새로운 문자열 xy도 VPS 가 된다. 예를 들어 “(())()”와 “((()))” 는 VPS 이지만 “(()(”, “(())()))” , 그리고 “(()” 는 모두 VPS 가 아닌 문자열이다. 

여러분은 입력으로 주어진 괄호 문자열이 VPS 인지 아닌지를 판단해서 그 결과를 YES 와 NO 로 나타내어야 한다. 

입력 데이터는 표준 입력을 사용한다. 입력은 T개의 테스트 데이터로 주어진다. 입력의 첫 번째 줄에는 입력 데이터의 수를 나타내는 정수 T가 주어진다. 각 테스트 데이터의 첫째 줄에는 괄호 문자열이 한 줄에 주어진다. 하나의 괄호 문자열의 길이는 2 이상 50 이하이다. 


어떤 자료구조를 사용해야 풀 수 있는 문제일까?

올바른 괄호쌍을 찾는 문제 : 스택을 이용하는 문제!

In [14]:
def VPS(data):
    stk = []
    for i in data:
        if i == '(':
            stk.append(i)
        else:
            if len(stk)== 0:
                return print("NO")
        stk.pop()
    if len(stk) == 0:
        return print("YES")
    else:
        return print("NO")
            


In [15]:
VPS('6')
VPS('(())())')
VPS('(((()())()')



NO
NO
NO


In [16]:
def is_vps(ps):
    stack = []  # 스택 초기화
    for p in ps:
        if p == '(':
            stack.append(p)
        else:
            if not stack:  # 스택이 비어있는 경우
                return print("N0")
            stack.pop()  # 짝이 맞는 경우 스택에서 꺼내기
    if not stack:  # 스택이 비어있는 경우
        return print("YES")
    else:
        return print("NO")

In [17]:
is_vps('6')
is_vps('(())())')
is_vps('(((()())()')

N0
N0
NO


In [18]:
# 입력받은 괄호 문자열
ps = input()

# 스택으로 사용할 빈 리스트
stack = []

# 올바른 괄호 문자열인지 판단하는 변수
valid = False

# 괄호 문자열을 왼쪽부터 오른쪽으로 순서대로 읽는다.
for p in ps:
    # 현재 읽은 문자가 '('이면 스택에 push한다.
    if p == '(':
        stack.append(p)
    # 현재 읽은 문자가 ')'이면 스택에서 pop을 시도한다.
    elif p == ')':
        # pop을 시도했을 때 스택이 비어있으면 올바르지 않은 괄호 문자열이다.
        if len(stack) == 0:
            valid = False
            break
        # pop을 시도했을 때 스택의 맨 위에 있는 데이터가 '('이면 pop을 성공적으로 수행한다.
        elif stack[-1] == '(':
            stack.pop()
        # pop을 시도했을 때 스택의 맨 위에 있는 데이터가 '('이 아니면 올바르지 않은 괄호 문자열이다.
        else:
            valid = False
            break

# 모든 문자를 읽고 난 후에 스택이 비어있으면 올바른 괄호 문자열이고, 비어있지 않으면 올바르지 않은 괄호 문자열이다.
    if len(stack) == 0 and valid == True:
        print("YES")
    else:
        print("NO")

NO
NO
NO
NO
NO
NO


In [19]:
for i in range(len(input())):
    stk = []
    isVPS = True
    for ch in input():
        if ch == '(':
            stk.append(ch)
        else:
            if stk:
                stk.pop()
            else:
                isVPS = False
                break
    if stk:
        isVPS = False

    print('YES' if isVPS else 'NO')
    break

YES


**2164 카드2**

N장의 카드가 있다. 각각의 카드는 차례로 1부터 N까지의 번호가 붙어 있으며, 1번 카드가 제일 위에, N번 카드가 제일 아래인 상태로 순서대로 카드가 놓여 있다.

이제 다음과 같은 동작을 카드가 한 장 남을 때까지 반복하게 된다. 우선, 제일 위에 있는 카드를 바닥에 버린다. 그 다음, 제일 위에 있는 카드를 제일 아래에 있는 카드 밑으로 옮긴다.

예를 들어 N=4인 경우를 생각해 보자. 카드는 제일 위에서부터 1234 의 순서로 놓여있다. 1을 버리면 234가 남는다. 여기서 2를 제일 아래로 옮기면 342가 된다. 3을 버리면 42가 되고, 4를 밑으로 옮기면 24가 된다. 마지막으로 2를 버리고 나면, 남는 카드는 4가 된다.

N이 주어졌을 때, 제일 마지막에 남게 되는 카드를 구하는 프로그램을 작성하시오.

---

처음들어온게 처음 나가므로 큐 자료구조를 생각할 수 있다


In [37]:
N = 500000
dack = []
# 카드 덱을 만든다
for i in range(N):
    dack.append(i+1)

for i in range(N-1):
# 맨 위의 카드를 제거
    del dack[0]
# 두번째 있던 카드를 맨 뒤에 붙여넣기
    dack.append(dack[0])
# 두번째 있던 카드 제거
    del dack[0]

dack

[475712]

삽입 삭제의 시간복잡도는 O(N) -> N-1번 반복 -> 총 시간 복잡도 O(N^2)

문제에서 주어진 최대 N = 500,000으로 계산해본 결과 시간이 35.8초 소요

큐의 삽입 삭제 시간복잡도는 O(1) -> N-1번 반복 -> 총 시간 복잡도 O(N)

큐를 사용했을 때에는 0초 소요

In [8]:
from collections import deque

N=500000
dq = deque(range(1,N+1))

while len(dq) > 1:
    dq.popleft()
    dq.append(dq.popleft())

dq[0]

475712

In [9]:
N=500000
dq = deque(range(1,N+1))

for i in range(N-1):
    dq.popleft()
    dq.append(dq.popleft())

dq[0]

475712

**11286. 절댓값 힙**

절댓값 힙은 다음과 같은 연산을 지원하는 자료구조이다.

배열에 정수 x (x ≠ 0)를 넣는다.
배열에서 절댓값이 가장 작은 값을 출력하고, 그 값을 배열에서 제거한다. 절댓값이 가장 작은 값이 여러개일 때는, 가장 작은 수를 출력하고, 그 값을 배열에서 제거한다.
프로그램은 처음에 비어있는 배열에서 시작하게 된다.

첫째 줄에 연산의 개수 N(1≤N≤100,000)이 주어진다. 다음 N개의 줄에는 연산에 대한 정보를 나타내는 정수 x가 주어진다. 만약 x가 0이 아니라면 배열에 x라는 값을 넣는(추가하는) 연산이고, x가 0이라면 배열에서 절댓값이 가장 작은 값을 출력하고 그 값을 배열에서 제거하는 경우이다. 입력되는 정수는 -231보다 크고, 231보다 작다.

---

힙에서는 최소값이 최상단에 위치하게 되므로 자료구조는 힙으로 구성할 수 있겠다

In [37]:
import heapq as hq

N = int(input())
pq = []
for _ in range(N):
    x = int(input())
    if x:
        hq.heappush(pq, (abs(x), x))
    else:
        if pq:
            print(hq.heappop(pq)[1])
        else:
            print(0)


-1
1
0
-1
-1
1
1
-2
-2


**1302. 베스트 셀러**

김형택은 탑문고의 직원이다. 김형택은 계산대에서 계산을 하는 직원이다. 김형택은 그날 근무가 끝난 후에, 오늘 판매한 책의 제목을 보면서 가장 많이 팔린 책의 제목을 칠판에 써놓는 일도 같이 하고 있다.

오늘 하루 동안 팔린 책의 제목이 입력으로 들어왔을 때, 가장 많이 팔린 책의 제목을 출력하는 프로그램을 작성하시오.

첫째 줄에 오늘 하루 동안 팔린 책의 개수 N이 주어진다. 이 값은 1,000보다 작거나 같은 자연수이다. 둘째부터 N개의 줄에 책의 제목이 입력으로 들어온다. 책의 제목의 길이는 50보다 작거나 같고, 알파벳 소문자로만 이루어져 있다.

---

무슨 자료구조? 

어떤 책이 몇번 팔렸다라는 정보를 저장해야겠지?
딕셔너리를 이용하는게 좋겠다!

In [46]:
d = dict()
N = int(input())

for _ in range(N):
    book = input()
    if book in d:
        d[book] += 1
    else:
        d[book] = 1

max_book = max(d.values())
candidate = []
for key, value in d.items():
    if value == max_book:
        candidate.append(key)

candidate.sort()
print(candidate[0])

a


자료구조
- array
- stack
- queue
- priority_queue
- dictionary
- set